### DISCUSSION TITLE
Adjacent matrix of PAG, from FCI, GFCI or RFCI in tetrad

### DISCUSSION THREAD

**Discussion by yasu-sh** [felixleopoldo/benchpress#80] [Q&A]
The manual says only about simple adjacent matrix as network structure DAG or CPDAG. The elements have only 0 or 1.
But FCI, GFCI and RFCI output originaly PAG. 
pcalg have the[ type definition in help](https://search.r-project.org/CRAN/refmans/pcalg/html/amatType.html).

The translation code from json in this repos looks treating o-> as ---, <-o as ---, o-o as ---. (i.e. CPDAG).
Is it true?

**Comment by felixleopoldo** [comment]
Thanks for pointing this out. I'll look it up. Though the FCI, RFCI, and GFCI are from causal-cmd so it should be that [docs](https://bd2kccd.github.io/docs/causal-cmd/) to look into.

**Reply by yasu-sh** [reply]
I appreciate it. I may help since I needed to read json at R code about a half year ago.
Finally we might have to  check the metrics of PAG in tetrad-gui..

**Reply by felixleopoldo** [reply]
Yes, the different edges are explained in the appendix [here](https://cmu-phil.github.io/tetrad/manual ). But maybe we can use the format from pcalg as it can be saved in CSV files. 
Yes, that could maybe be part of the benchmarks module.

**Reply by yasu-sh** [reply]
Thanks for checking them. TETRAD GUI have pcalg export function. As I checked, the rule works fine.
![image](https://github.com/felixleopoldo/benchpress/assets/89838961/2688ce2c-b9d0-4132-9c5d-cd47821554e8)
If  the method for pcalg export is available, it would be good for us.

Code of export in tetrad-lib
https://github.com/cmu-phil/tetrad/blob/1b696a2ecd92bfcd64fd44658b7b8804eb6cebe5/tetrad-lib/src/main/java/edu/cmu/tetrad/graph/GraphSaveLoadUtils.java#L769

**Reply by felixleopoldo** [reply]
Ah, that looks promising, maybe there is a way to do this from causal-cmd too.

**Reply by felixleopoldo** [reply]
In the future, we might have to change the adjacency matrix format to edge lists, as the graphs treated are usually rather sparse and the matrices don't really scale well. But let's save that for a future task.

**Reply by yasu-sh** [reply]
Yes. I agree. Now python users have two main ways to access to tetrad-java, i.e. causal-cmd and py-tetrad.
py-tetrad have direct access to jar file and enables jar to get pcalg matrix format. (I have not seen normal adjacent matrix yet.)
![image](https://github.com/felixleopoldo/benchpress/assets/89838961/f3ecd065-e332-4d70-809d-2d681805488f)

I think benchpress has R code for transform the egdelist to adjmatrix. both causal-cmd and py-tetrad are good tools.
As future tasks, let's find better way.

**Comment by felixleopoldo** [comment]
The PAG algorithms are moved to the PAGs branch for now until we resolve this.

**Reply by yasu-sh** [reply]
It's reasonable choice. 
I should have noted that pcalg's format(a --> b means m[b, a]) is tranposed as bnlearn's one(a --> b means m[a, b]).  
The py-tetrad example above uses pcalg's one.
I would like to make an issue to add [some notes on it.](https://benchpressdocs.readthedocs.io/en/latest/data_formats.html#graph) Hopefully you can do it. 

[Two definition types](https://en.wikipedia.org/wiki/Adjacency_matrix#:~:text=%E6%9C%89%E5%90%91%E3%82%B0%E3%83%A9%E3%83%95%E3%81%AE%E9%9A%A3%E6%8E%A5%E8%A1%8C%E5%88%97%E3%81%AF%E3%80%81%E6%AC%A1%E3%81%AE%E3%81%84%E3%81%9A%E3%82%8C,i%E3%81%B8%E3%81%AE%E3%82%A8%E3%83%83%E3%82%B8%E3%82%92%E7%A4%BA%E3%81%97%E3%81%BE%E3%81%99%E3%80%82)

**Reply by yasu-sh** [reply]
The reason above is: 

- I needed to check ./resources/adjmat/myadjmats/alarm.csv and confirmed the same contents with https://www.bnlearn.com/bnrepository/
- I needed to check pcalg homepage https://search.r-project.org/CRAN/refmans/pcalg/html/amatType.html and check some commands try many types: i.e.  

```
> library(pcalg)
> data("gmD")
> V <- colnames(gmD$x)
> ## define sufficient statistics
> suffStat <- list(dm = gmD$x, nlev = c(3,2,3,4,2), adaptDF = FALSE)
> ## estimate CPDAG
> pc.D <- pc(suffStat,
+            ## independence test: G^2 statistic
+            indepTest = disCItest, alpha = 0.01, labels = V)
> as(pc.D,"amat")
Adjacency Matrix 'amat' (5 x 5) of type ‘cpdag’:
   X1 X2 X3 X4 X5
X1  .  .  .  .  .
X2  .  .  .  .  .
X3  1  1  .  .  .
X4  .  1  .  .  1
X5  .  .  .  .  .
> as(pc.D@graph,"matrix")
   X1 X2 X3 X4 X5
X1  0  0  1  0  0
X2  0  0  1  1  0
X3  0  0  0  0  0
X4  0  0  0  0  0
X5  0  0  0  1  0
> summary(pc.D)
Object of class 'pcAlgo', from Call:
pc(suffStat = suffStat, indepTest = disCItest, alpha = 0.01, 
    labels = V)

Nmb. edgetests during skeleton estimation:
===========================================
Max. order of algorithm:  1 
Number of edgetests from m = 0 up to m = 1 :  14 6

Graphical properties of skeleton:
=================================
Max. number of neighbours:  2 at node(s) 2 
Avg. number of neighbours:  0.8 

Adjacency Matrix G:
   X1 X2 X3 X4 X5
X1  .  .  1  .  .
X2  .  .  1  1  .
X3  .  .  .  .  .
X4  .  .  .  .  .
X5  .  .  .  1  .
```

I think users would be able to understand when they look into intermediate files to own data/code debug.

**Reply by yasu-sh** [reply]
let me unmark as answered, it is useful and I may quote in near future.
